### DISCUSSION TITLE
Allow opening hyperlinks

### DISCUSSION THREAD

**Discussion by ghost** [netblue30/firejail#5566] [General]
With the upcoming 0.9.72 release approaching I'm going over some profiles to check a few things that have been in my notes for a while. One of these topics is (how to) allow opening hyperlinks. Here are my observations/questions that came up in that context.

The hyperlink support functionality seems to rely on the following firejail options being present:
- `access to Firefox profile info`
  - noblacklist ${HOME}/.mozilla
  - whitelist ${HOME}/.mozilla/firefox/profiles.ini
  - read-only ${HOME}/.mozilla/firefox/profiles.ini
- `access to the Firefox D-Bus user bus`
  - dbus-user.talk org.mozilla.*
- `access to a shell`
  - include allow-bin-sh.inc when disable-shell.inc is present
  - bash,sh being present in private-bin when that is used

If those are indeed _all_ needed I think we should double-check/test our profiles that currently have support for opening hyperlinks. The list below includes a request for checking/confirming this functionality, adressed at the original creators of the profile.

`balsa.profile` @kortewegdevries [commit](https://github.com/netblue30/firejail/commit/7df28c1edeb50f6b856331ba65da05afae7ae81f)
- include allow-bin-sh.inc is missing
- bash,sh missing from private-bin
- dbus-user.talk org.mozilla.* is not mentioned in the D-Bus section

`kube.profile` @kortewegdevries [commit](https://github.com/netblue30/firejail/commit/7df28c1edeb50f6b856331ba65da05afae7ae81f)
- include allow-bin-sh.inc is missing
- bash,sh missing from private-bin
- dbus-user.talk org.mozilla.* is not mentioned in the D-Bus section

`telegram.profile` @BaderSZ [commit](https://github.com/netblue30/firejail/commit/f7598b95677f014b5ce3696b57c5ed25d4a92b1c)
- access to Firefox profile info is missing
  - noblacklist ${HOME}/.mozilla
  - whitelist ${HOME}/.mozilla/firefox/profiles.ini
  - read-only ${HOME}/.mozilla/firefox/profiles.ini

`trojita` @kortewegdevries [commit](https://github.com/netblue30/firejail/commit/7df28c1edeb50f6b856331ba65da05afae7ae81f)
- include allow-bin-sh.inc is missing
- bash,sh missing from private-bin
- dbus-user.talk org.mozilla.* is not mentioned in the D-Bus section

`zeal.profile` @rusty-snake [commit](https://github.com/netblue30/firejail/commit/b3967fade379b5b6a073df3d42574455314b6db6)
- include allow-bin-sh.inc is missing
- bash,sh missing from private-bin

Although the original profile creators have been referenced, anyone who can confirm the hyperlink functionality in the above profiles is working as expected is asked to give a shout.

I've already opened PR's for profiles I created/use daily and which needed fixes. Feel free to comment on those as well:

- #5564
- #5565

Let's get this all in shape for 0.9.72, TIA!

Relates to:

- #6462

**Comment by rusty-snake** [comment]
@bbhtt see OP.

---

IMHO we should fix the "Firejail URL open game" (#5364) instead of adding level 1 hack to more profiles and keep maintaining it working.

**Reply by ghost** [reply]
**I totally agree**. FWIW level 3 & 4 look the better options to me. Better, as in easier to integrate into Firejail. That being said, if we cannot fix this "Firejail URL open game" before releasing 0.9.72, some profiles would be left in a broken state regarding hyperlinking functionality. I'm `not` suggesting here to add level 1 hack to `more` profiles but to `fix what's already out there`.

There's also [firefox-socket](https://github.com/karabaja4/firefox-socket-control) which is very similar to [fireurl](https://github.com/rusty-snake/fireurl). Controlling Firefox via a UNIX socket is awesome :-)

Just a question. What do you think is needed to be able to get a proper fix into 0.9.72, if possible? Do you suggest to drop the level 1 hacks we have now and advise users to take care of the URL opening by themselves via [xdg-open.c](https://gist.github.com/rusty-snake/5104dc53ce3e52eef86cc34d359aa10e), [fireurl](https://github.com/rusty-snake/fireurl) or whatever they see fit?

**Reply by bbhtt** [reply]
> @bbhtt see OP.

From my testing for Firefox stable installed through pacman, set as default browser through `xdg-mime`, `whitelist ${HOME}/.mozilla/firefox/profiles.ini` doesn't work, Firefox launches but reports "no profile found". 

`mkdir ${HOME}/.mozilla/firefox` and `whitelist ${HOME}/.mozilla/firefox` seems to work.

`seccomp` makes Firefox unresponsive.

<details><summary>Balsa</summary>

```diff
diff --git a/etc/profile-a-l/balsa.profile b/etc/profile-a-l/balsa.profile
index 661356ff6..d6326b477 100644
--- a/etc/profile-a-l/balsa.profile
+++ b/etc/profile-a-l/balsa.profile
@@ -14,6 +14,8 @@ noblacklist ${HOME}/mail
 noblacklist /var/mail
 noblacklist /var/spool/mail
 
+include allow-bin-sh.inc
+
 include disable-common.inc
 include disable-devel.inc
 include disable-exec.inc
@@ -24,11 +26,12 @@ include disable-xdg.inc
 
 mkdir ${HOME}/.balsa
 mkdir ${HOME}/.gnupg
+mkdir ${HOME}/.mozilla/firefox
 mkfile ${HOME}/.signature
 mkdir ${HOME}/mail
 whitelist ${HOME}/.balsa
 whitelist ${HOME}/.gnupg
-whitelist ${HOME}/.mozilla/firefox/profiles.ini
+whitelist ${HOME}/.mozilla/firefox
 whitelist ${HOME}/.signature
 whitelist ${HOME}/mail
 whitelist ${RUNUSER}/gnupg
@@ -56,13 +59,13 @@ notv
 nou2f
 novideo
 protocol unix,inet,inet6
-seccomp
+#seccomp
 tracelog
 
 # disable-mnt
 # Add "pinentry-curses,pinentry-emacs,pinentry-fltk,pinentry-gnome3,pinentry-gtk,pinentry-gtk2,pinentry-gtk-2,pinentry-qt,pinentry-qt4,pinentry-tty,pinentry-x2go,pinentry-kwallet" for gpg
 # Add "ignore private-bin" for hyperlinks or have a look at the private-bins in firefox.profile and firefox-common.profile.
-private-bin balsa,balsa-ab,gpg,gpg-agent,gpg2,gpgsm
+#private-bin balsa,balsa-ab,gpg,gpg-agent,gpg2,gpgsm
 private-cache
 private-dev
 private-etc alternatives,ca-certificates,crypto-policies,dconf,fonts,gcrypt,groups,gtk-2.0,gtk-3.0,hostname,hosts,ld.so.cache,ld.so.preload,mailname,passwd,pki,resolv.conf,selinux,ssl,xdg
@@ -76,7 +79,8 @@ dbus-user.talk ca.desrt.dconf
 dbus-user.talk org.freedesktop.Notifications
 dbus-user.talk org.freedesktop.secrets
 dbus-user.talk org.gnome.keyring.SystemPrompter
+dbus-user.talk org.mozilla.*
 dbus-system none
```
</details>

This setup breaks easily:  a different browser is used, Firefox installed as a flatpak, Firefox is installed from the mozilla tarball (placed under `/opt` or `~/.local/share` etc.)

<details><summary>PS</summary>

I think these email clients should redirect to `email-common.profile`

</details>

**Reply by ghost** [reply]
@bbhtt

Thanks for testing! Your observations do confirm the current `level 1 hack` as @rusty-snake refers to in #5364 sure needs a proper fix. Not sure if we can do that in the upcoming 0.9.72 though.

> I think these email clients should redirect to email-common.profile

Yeah we've had similar remarks about our email clients. I'll have another look at this, should be feasible to come up with a better strategy indeed.

**Reply by ghost** [reply]
@rusty-snake I'll be looking into creating an AUR package for your [fireurl](https://github.com/rusty-snake/fireurl). That seems the best option right now to puth these shaky level 1 hacks to rest. Is there a way Fedora users can get that via rpm? Perhaps @reinerh can assist here for the Debian/Ubuntu side?

**Reply by rusty-snake** [reply]
I can create a spec (PKGBUILD for rpm) for it. Or just two static linked binaries that should work everywhere.

**Reply by ghost** [reply]
Lovely. Two statically linked binaries should be fine for users to easily get `fireurl` regardless of distribution. One more question. I will be testing fireurl shortly, but if you could provide info on what profiles need to get it working properly that would save me some time and I could start work on cleaning up level 1 hack stuff. Is filesystem access to Firefox profile info still needed?

**Reply by rusty-snake** [reply]
Only `${RUNUSER}/fireurl` and the obivious stuff (binary itself, dynamic linker and libc, ...) should be required. And `unix` in `protocol` as well as all used syscalls. Maybe mdwe is a Problem IDK.

TL;DR: `whitelist ${RUNUSER}/fireurl` in `whitelist-runuser-common.inc` should be enougth for IDK ~70% of our profiles. For the next 35% you need `noblacklist ${RUNUSER}` or `private-bin fireurl`.

**Reply by ghost** [reply]
Thanks for the info. All makes sense so I can work with that!

**Reply by rusty-snake** [reply]
I made a 0.1.0 release of fireurl: https://github.com/rusty-snake/fireurl/releases/tag/v0.1.0

**Reply by ghost** [reply]
Nice work on the static binaries! In the mean time I went over our profiles and counted only 10 profiles that have 'level 1 hack' support at the time of writing. They all happen to include `wurc` so porting them to support fireurl would be a breeze. Not every app we have profiles for would need the hyperlink functionality anyway. I have everything needed to prep a PR for that.

My only remaining question is why don't we add these two small executables to `contrib` (together with the fireurld systemd unit) and make life easy for firejail package maintainers? Adding a classic fireurld.desktop XDG-autostart file for both systemd/non-systemd OS'es is another option:

```console
$ cat /etc/xdg/autostart/fireurld.desktop
[Desktop Entry]
Name=FireURL Daemon
Comment=Open hyperlinks between Firejail sandboxes via UNIX domain socket
Exec=fireurld
Icon=fireurld
StartupNotify=false
Type=Application

NoDisplay=true
X-GNOME-HiddenUnderSystemd=true
X-GNOME-UsesNotifications=false
X-systemd-skip=true

```

Maybe I'm over-simplifying this fireurl integration into firejail. That would be a decision both @netblue30 and you need to agree upon I guess...

**Reply by rusty-snake** [reply]
>Maybe I'm over-simplifying this fireurl integration into firejail.

Just shipping fireurl with firejail is simple. If you however want to integrate fireurl into firejail by starting fireurld if it is not running and mounting the socket at e.g. `/run/firejail/fireurl.sock` i.e. making fireurl always working w/o profile changes or anything else, requires more work.

**Reply by ghost** [reply]
> If you however want to integrate fireurl into firejail by starting fireurld if it is not running and mounting the socket at e.g. /run/firejail/fireurl.sock

Sure. One more fireurl.desktop file, placed in /usr/share/applications for example and (informing the user about) setting that as the default web browser should do it correct? I'm pretty sure package managers can deal with all this.

```console
$ cat /usr/share/applications/fireurl.desktop
[Desktop Entry]
Name=FireURL Client
Comment=Open hyperlinks between Firejail sandboxes via UNIX domain socket
Exec=fireurl %U
Icon=fireurld
Type=Application
Categories=Network;WebBrowser;

MimeType=text/html;x-scheme-handler/http;x-scheme-handler/https;

```

**Reply by ghost** [reply]
To clarify my opinion: pushing for a solution to be 100% ready for the upcoming 0.9.72 is not what's important. Ofcourse we need to think this through and have an open discussion on. That being said, the recurring questions/issues about this functionality go back years. Sometimes one needs to break a few eggs if one wants to make an omelet...

**Comment by chrisgoldberg1** [comment]
Has Fireurl made it into Firejail now, or how many more years will this annoying issue actually remain a problem? How do you deal with situations like opening links in Telegram or Thunderbird? Do we really still have to copy and paste links in 2025 when using Firejail? I've encountered this issue several times and always searched for a solution, but it seems there's no proper solution that doesn't involve significant "hacks."
